sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -572,6 +572,10 @@ class Dataset[T] private[sql](
     val parsedDelay =
       Option(CalendarInterval.fromString("interval " + delayThreshold))
         .getOrElse(throw new AnalysisException(s"Unable to parse time delay '$delayThreshold'"))
+    // Threshold specified in months/years is non-deterministic
```

Comment:
I think waiting 1 month for late data is a reasonable use case. Based on the definition of the watermark, its actually okay for us to over estimate this delay too.  Why not take take the max (31 days, leap year)?